crates/sdk/src/masp.rs

```diff
@@ -1621,235 +1609,309 @@ impl<U: ShieldedUtils + MaybeSend + MaybeSync> ShieldedContext<U> {
             // use from the masp crate to specify the expiration better
             expiration_height.into(),
         );
+        // Determine epoch in which to submit potential shielded transaction
+        let epoch = rpc::query_masp_epoch(context.client()).await?;
 
-        // Convert transaction amount into MASP types
-        let Some(denom) = query_denom(context.client(), token).await else {
-            return Err(TransferErr::General(Error::from(
-                QueryError::General(format!("denomination for token {token}")),
-            )));
-        };
-        let (asset_types, masp_amount) = {
-            let mut shielded = context.shielded_mut().await;
-            // Do the actual conversion to an asset type
-            let amount = shielded
-                .convert_amount(
-                    context.client(),
-                    epoch,
-                    token,
-                    denom,
-                    amount.amount(),
-                )
-                .await?;
-            // Make sure to save any decodings of the asset types used so that
-            // balance queries involving them are successful
-            let _ = shielded.save().await;
-            amount
-        };
-
-        // If there are shielded inputs
-        if let Some(sk) = spending_key {
-            // Locate unspent notes that can help us meet the transaction amount
-            let (_, unspent_notes, used_convs) = context
-                .shielded_mut()
-                .await
-                .collect_unspent_notes(
-                    context,
-                    &to_viewing_key(&sk).vk,
-                    I128Sum::from_sum(masp_amount),
-                    epoch,
-                )
-                .await?;
-            // Commit the notes found to our transaction
-            for (diversifier, note, merkle_path) in unspent_notes {
-                builder
-                    .add_sapling_spend(sk, diversifier, note, merkle_path)
-                    .map_err(builder::Error::SaplingBuild)?;
-            }
-            // Commit the conversion notes used during summation
-            for (conv, wit, value) in used_convs.values() {
-                if value.is_positive() {
-                    builder
-                        .add_sapling_convert(
-                            conv.clone(),
-                            *value as u64,
-                            wit.clone(),
-                        )
-                        .map_err(builder::Error::SaplingBuild)?;
-                }
+        let mut is_context_loaded = false;
+        for MaspTransferData {
+            source,
+            target,
+            token,
+            amount,
+        } in data
+        {
```

Comment:
Just a note: I think I've improved this in #3393 by trying to group the same sources/tokens before looking for the available notes